sdk/src/azure/core/az_mqtt5_rpc_client_hfsm.c

```diff
@@ -0,0 +1,691 @@
+  az_span status_str = az_mqtt5_property_stringpair_get_value(status);
+  if (az_result_failed((az_span_atoi32(status_str, (int32_t*)&resp_data.status))))
+  {
+    resp_data.parsing_failure = true;
+    resp_data.error_message = AZ_SPAN_FROM_STR("Status property contains invalid value.");
+    return resp_data;
+  }
+
+  if (resp_data.status < 200 || resp_data.status >= 300)
+  {
+    // read the error message if there is one
+    // az_mqtt5_property_stringpair error_message;
+    if (!az_result_failed(az_mqtt5_property_bag_find_stringpair(
+            recv_data->properties,
+            AZ_MQTT5_PROPERTY_TYPE_USER_PROPERTY,
+            AZ_SPAN_FROM_STR(AZ_MQTT5_RPC_STATUS_MESSAGE_PROPERTY_NAME),
+            error_message)))
+    {
+      resp_data.error_message = az_mqtt5_property_stringpair_get_value(error_message);
+    }
+  }
+  else
+  {
+    if (az_span_is_content_equal(recv_data->payload, AZ_SPAN_EMPTY))
+    {
+      resp_data.parsing_failure = true;
+      resp_data.error_message = AZ_SPAN_FROM_STR("Empty payload in 200");
+      return resp_data;
+    }
+    // read the content type so the application can properly deserialize the request
+    // az_mqtt5_property_string content_type;
+    if (!az_result_failed(az_mqtt5_property_bag_read_string(
+            recv_data->properties, AZ_MQTT5_PROPERTY_TYPE_CONTENT_TYPE, content_type)))
+    {
+      resp_data.content_type = az_mqtt5_property_get_string(content_type);
+    }
+  }
+
+  return resp_data;
+}
+
+static az_result root(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  az_mqtt5_rpc_client_hfsm* this_policy = (az_mqtt5_rpc_client_hfsm*)me;
+
+  if (_az_LOG_SHOULD_WRITE(event.type))
+  {
+    _az_LOG_WRITE(event.type, AZ_SPAN_FROM_STR("az_rpc_client_hfsm"));
+  }
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      break;
+
+    case AZ_HFSM_EVENT_ERROR:
+    {
+      _az_RETURN_IF_FAILED(_az_mqtt5_connection_api_callback(
+          this_policy->_internal.connection,
+          (az_event){ .type = AZ_HFSM_EVENT_ERROR, .data = event.data }));
+      // TODO: uncomment when this won't fail every time
+      // if (az_result_failed(az_event_policy_send_inbound_event(me, event)))
+      // {
+      //   az_platform_critical_error();
+      // }
+      break;
+    }
+
+    case AZ_HFSM_EVENT_EXIT:
+    {
+      if (_az_LOG_SHOULD_WRITE(AZ_HFSM_EVENT_EXIT))
+      {
+        _az_LOG_WRITE(AZ_HFSM_EVENT_EXIT, AZ_SPAN_FROM_STR("az_mqtt5_rpc_client_hfsm: PANIC!"));
+      }
+
+      az_platform_critical_error();
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_PUBACK_RSP:
+    case AZ_MQTT5_EVENT_SUBACK_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_OPEN_REQ:
+    case AZ_MQTT5_EVENT_CONNECT_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_CLOSE_REQ:
+    case AZ_MQTT5_EVENT_DISCONNECT_RSP:
+    case AZ_MQTT5_EVENT_UNSUBACK_RSP:
+      break;
+
+    case AZ_EVENT_RPC_CLIENT_UNSUB_REQ:
+    {
+      // Send unsubscribe
+      az_mqtt5_unsub_data unsubscription_data
```

Comment:
I think all actions should be extracted as inline functions instead of detailing the operation here.